core/kroma_migration.go

```diff
@@ -0,0 +1,200 @@
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	return mr.root
+}
+
+func (mr *MigratedRef) BlockNumber() uint64 {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	return mr.number
+}
+
+// encodeBlockNumber encodes a block number as big endian uint64
+func encodeBlockNumber(number uint64) []byte {
+	enc := make([]byte, 8)
+	binary.BigEndian.PutUint64(enc, number)
+	return enc
+}
+
+func AccountChangesKey(blockNumber uint64) []byte {
+	return append(accountChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func StorageChangesKey(blockNumber uint64) []byte {
+	return append(storageChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func DestructChangesKey(blockNumber uint64) []byte {
+	return append(destructChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func ReadStateChanges(db ethdb.KeyValueStore, blockNumber uint64) (map[common.Address]bool, map[common.Hash][]byte, map[common.Hash]map[common.Hash][]byte, error) {
+	enc, err := db.Get(DestructChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	destruct, err := DeserializeStateChanges[map[common.Address]bool](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	enc, err = db.Get(AccountChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	accounts, err := DeserializeStateChanges[map[common.Hash][]byte](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	enc, err = db.Get(StorageChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	storages, err := DeserializeStateChanges[map[common.Hash]map[common.Hash][]byte](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return destruct, accounts, storages, nil
+}
+
+// Note: it's should be called in MigrationTime
+func WriteStateChanges(db ethdb.KeyValueStore, blockNumber uint64, stateObjectsDestruct map[common.Address]*types.StateAccount, accounts map[common.Hash][]byte, storages map[common.Hash]map[common.Hash][]byte) error {
+	batch := db.NewBatch()
+	stateObjectsDestructInfo := make(map[common.Address]bool)
+
+	// TODO: Do we need to check if batch.ValueSize() > ethdb.IdealBatchSize, have storages size limit?
+	for addr := range stateObjectsDestruct {
+		if addr.Cmp(params.SystemAddress) == 0 {
+			continue
+		}
+		stateObjectsDestructInfo[addr] = true
+
+	}
+
+	serializedDestruct, err := SerializeStateChanges(stateObjectsDestructInfo)
+	if err != nil {
+		return err
+	}
+	err = batch.Put(DestructChangesKey(blockNumber), serializedDestruct)
+	if err != nil {
+		return err
+	}
+
+	serializedAccounts, err := SerializeStateChanges(accounts)
+	if err != nil {
+		return err
+	}
+	err = batch.Put(AccountChangesKey(blockNumber), serializedAccounts)
+	if err != nil {
+		return err
+	}
+	serializedStorages, err := SerializeStateChanges(storages)
+	if err != nil {
+		return err
+	}
+	err = batch.Put(StorageChangesKey(blockNumber), serializedStorages)
+	if err != nil {
+		return err
+	}
+	if err := batch.Write(); err != nil {
+		return err
+	}
+	batch.Reset()
```

Comment:
batch seems to be written only within this block. Do we need a reset?
```suggestion
	
```

Reply:
I agree with your point. I'll update the code to remove the reset, as it seems unnecessary here.